File: GravePlan/src/main/java/de/hannisoft/gaveplan/excelimport/GraveSiteFileReader.java
package de.hannisoft.gaveplan.excelimport;

import java.io.File;
import java.io.IOException;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Map;
import java.util.TreeMap;

import de.hannisoft.gaveplan.model.GraveSite;
import de.hannisoft.gaveplan.model.GraveSiteType;
import de.hannisoft.gaveplan.model.Owner;
import jxl.Sheet;
import jxl.Workbook;
import jxl.WorkbookSettings;
import jxl.read.biff.BiffException;

public class GraveSiteFileReader extends AbstractXLSFileReader {
    private static final String COL_FIELD = "Feld";
    private static final String COL_ROW = "Reihe";
    private static final String COL_PLACE = "Grabstätte";
    private static final String COL_TYPE = "Grabart";
    private static final String COL_NAME = "Grabname";
    private static final String COL_VALID_FROM = "Nutzungsbeginn";
    private static final String COL_VALID_TO = "Nutzungsende";
    private static final String COL_SIZE = "Grabstellenanzahl";
    private static final String COL_LAST_NAME = "Nachname";
    private static final String COL_FIRST_NAME = "Vorname";
    private static final String COL_STREET = "Straße";
    private static final String COL_BUILDING_NO = "Hausnr.";
    private static final String COL_ZIP = "PLZ";
    private static final String COL_CITY = "Ort";
    private static final String COL_SALUTAION = "Anrede";
    private static final String COL_SALUTATION_LETTER = "Briefanrede";

    public Map<String, GraveSite> read(String inputFile) throws IOException {
        Map<String, GraveSite> graves = new TreeMap<>();
        File inputWorkbook = new File(inputFile);
        Workbook w;
        WorkbookSettings ws = new WorkbookSettings();
        ws.setEncoding("Cp1252");
        try {
            w = Workbook.getWorkbook(inputWorkbook, ws);
            Sheet sheet = w.getSheet(0);
            initColumnMap(sheet, 0);

            // DateFormat format = new SimpleDateFormat("dd.MM.yyyy");
            DateFormat format = new SimpleDateFormat("dd/MM/yyyy");
            for (int i = 1; i < sheet.getRows(); i++) {
                String field = getContent(COL_FIELD, i);
                if (field == null || field.trim().isEmpty()) {
                    continue;
                }
                String row = getContent(COL_ROW, i);
                String place = getContent(COL_PLACE, i);
                String type = getContent(COL_TYPE, i);
                String name = getContent(COL_NAME, i);
                String validFrom = getContent(COL_VALID_FROM, i);
                String validTo = getContent(COL_VALID_TO, i);
                String size = getContent(COL_SIZE, i);

                Owner owner = new Owner();
                owner.setLastName(getContent(COL_LAST_NAME, i));
                owner.setFirstName(getContent(COL_FIRST_NAME, i));
                owner.setStreet(getContent(COL_STREET, i) + " " + getContent(COL_BUILDING_NO, i));
                owner.setZipAndTown(getContent(COL_ZIP, i) + " " + getContent(COL_CITY, i));
                owner.setSaluation(getContent(COL_SALUTAION, i));
                owner.setSaluationLetter(getContent(COL_SALUTATION_LETTER, i));

                try {
                    GraveSite graveSite = new GraveSite(field, row, place);
                    graveSite.setType(GraveSiteType.getTypeByName(type));
                    graveSite.setOwner(owner);
                    graveSite.setName(name);
                    if (validFrom != "") {
                        graveSite.setValidFrom(format.parse(validFrom));
                    }
                    if (validTo != "") {
                        graveSite.setValidTo(format.parse(validTo));
                    }
                    graveSite.setSize(Integer.parseInt(size));
                    graves.put(graveSite.getId(), graveSite);
                } catch (Exception e) {
                    System.err.println("Invalid GraveSite: " + field + "/" + row + "/" + place);
                    e.printStackTrace();
                }
            }
        } catch (BiffException e) {
            e.printStackTrace();
            return null;
        }
        return graves;
    }

}

File: src/main/kotlin/de/hannisoft/graveplan/excelimport/GraveSiteFileReader.kt
package de.hannisoft.de.hannisoft.graveplan.excelimport

import de.hannisoft.graveplan.model.GraveSite
import de.hannisoft.graveplan.model.Owner
import org.apache.poi.ss.usermodel.WorkbookFactory
import java.io.File
import java.text.DateFormat
import java.text.SimpleDateFormat

class GraveSiteFileReader() {
    companion object {
        const val COL_FIELD: String = "Feld"
        const val COL_ROW: String = "Reihe"
        const val COL_PLACE: String = "Grabstätte"
        const val COL_TYPE: String = "Grabart"
        const val COL_NAME: String = "Grabname"
        const val COL_VALID_FROM: String = "Nutzungsbeginn"
        const val COL_VALID_TO: String = "Nutzungsende"
        const val COL_SIZE: String = "Grabstellenanzahl"
        const val COL_LAST_NAME: String = "Nachname"
        const val COL_FIRST_NAME: String = "Vorname"
        const val COL_STREET: String = "Straße"
        const val COL_BUILDING_NO: String = "Hausnr."
        const val COL_ZIP: String = "PLZ"
        const val COL_CITY: String = "Ort"
        const val COL_SALUTAION: String = "Anrede"
        const val COL_SALUTATION_LETTER: String = "Briefanrede"
    }

    fun read(inputFile: File): Map<String, GraveSite> {
        val graves = mutableMapOf<String, GraveSite>()
        val format: DateFormat = SimpleDateFormat("dd/MM/yyyy")

        inputFile.inputStream().use { fis ->
            val workbook = WorkbookFactory.create(fis)
            val sheet = workbook.getSheetAt(0)
            val headerMap = sheet.buildHeaderMap()

            for (rowIndex in 1..sheet.lastRowNum) {
                val row = sheet.getRow(rowIndex) ?: continue

                val field = row.getValue(headerMap[COL_FIELD])
                if (field.isEmpty()) {
                    continue
                }

                val rowStr = row.getValue(headerMap[COL_ROW])
                val placeStr = row.getValue(headerMap[COL_PLACE])
                val type = row.getValue(headerMap[COL_TYPE])
                val name = row.getValue(headerMap[COL_NAME])
                val validFrom = row.getValue(headerMap[COL_VALID_FROM])
                val validTo = row.getValue(headerMap[COL_VALID_TO])
                val size = row.getValue(headerMap[COL_SIZE])

                val owner = Owner(row.getValue(headerMap[COL_SALUTAION]),
                    row.getValue(headerMap[COL_SALUTATION_LETTER]),
                    row.getValue(headerMap[COL_FIRST_NAME]),
                    row.getValue(headerMap[COL_LAST_NAME]),
                    row.getValue(headerMap[COL_STREET])+ " " + row.getValue(headerMap[COL_BUILDING_NO]),
                    row.getValue(headerMap[COL_ZIP])+ " " + row.getValue(headerMap[COL_CITY]))
                val graveSite = GraveSite(field, rowStr, placeStr)
                graveSite.typeAsString = type
                graveSite.owner = owner
                graveSite.name = name
                if (validFrom.isNotEmpty()) {
                    graveSite.validFrom = format.parse(validFrom)
                }
                if (validTo.isNotEmpty()) {
                    graveSite.validTo = format.parse(validTo)
                }
                graveSite.size = size.toInt()
                graves[graveSite.id] = graveSite
            }
        }
        return graves
    }
}